src/material/button/button.ts

```diff
@@ -97,21 +111,27 @@ export class MatButton extends _MatButtonMixinBase
 
   constructor(elementRef: ElementRef,
               private _focusMonitor: FocusMonitor,
-              @Optional() @Inject(ANIMATION_MODULE_TYPE) public _animationMode: string) {
+              @Optional() @Inject(ANIMATION_MODULE_TYPE) public _animationMode: string,
+              @Optional() @Inject(MAT_BUTTON_DEFAULT_OPTIONS)
+                  defaultOptions?: MatButtonDefaultOptions) {
     super(elementRef);
 
+    const hostElement = this._getHostElement();
+
+    if (!this._hasHostAttributes('type')) {
+      hostElement.type = defaultOptions?.type ?? 'submit';
+    }
+
     // For each of the variant selectors that is present in the button's host
     // attributes, add the correct corresponding class.
-    for (const attr of BUTTON_HOST_ATTRIBUTES) {
-      if (this._hasHostAttributes(attr)) {
-        (this._getHostElement() as HTMLElement).classList.add(attr);
-      }
-    }
+    BUTTON_HOST_ATTRIBUTES
+        .filter(attr => this._hasHostAttributes(attr))
```

Comment:
This is slightly nicer to look at, but it'll take up more memory since it creates an extra array. It was written with a plain loop before since apps can have lots of buttons on the page at the same time.

Reply:
Reverted.